src/guardian/staking-module-guard/staking-module-guard.service.ts

```diff
@@ -30,34 +30,85 @@ export class StakingModuleGuardService {
     {};
 
-  public async getStakingRouterModuleData(
-    stakingRouterModule: SRModule,
-    blockHash: string,
-  ): Promise<StakingModuleData> {
-    const {
-      data: {
-        keys,
-        module: { nonce },
-      },
-    } = await this.stakingRouterService.getStakingModuleUnusedKeys(
-      blockHash,
-      stakingRouterModule,
-    );
+  /**
+   * @returns List of staking modules id with duplicates
+   */
+  public checkVettedKeysDuplicates(
+    stakingModulesData: StakingModuleData[],
+    blockData: BlockData,
+  ): number[] {
+    // Collects the duplicate count for each unique key across staking modules.
+    // The outer Map uses the key string as the key and holds an inner Map.
+    // The inner Map uses module id as keys and stores the duplicate count for each module.
+    const keyMap = new Map<string, Map<number, number>>();
+    const modulesWithDuplicatedKeysSet = new Set<number>();
+    const duplicatedKeys = new Map<string, Map<number, number>>();
+
+    stakingModulesData.forEach(({ vettedKeys, stakingModuleId }) => {
+      // check module keys on duplicates across all modules
+      vettedKeys.forEach((key) => {
+        const stakingModules = keyMap.get(key.key);
+
+        if (!stakingModules) {
+          // add new key
+          keyMap.set(key.key, new Map([[stakingModuleId, 1]]));
+        } else {
+          // found duplicate
+          // Duplicate key found
+          const moduleCount = stakingModules.get(stakingModuleId) || 0;
+          stakingModules.set(stakingModuleId, moduleCount + 1);
+
+          if (this.hasDuplicateKeys(stakingModules)) {
+            stakingModules.forEach((_, id) => {
+              modulesWithDuplicatedKeysSet.add(id);
+            });
+            duplicatedKeys.set(key.key, stakingModules);
+          }
+        }
+      });
+    });
 
-    const isDepositsPaused = await this.securityService.isDepositsPaused(
-      stakingRouterModule.id,
-      {
-        blockHash,
-      },
-    );
+    if (modulesWithDuplicatedKeysSet.size) {
+      const moduleAddressesWithDuplicatesList: number[] = Array.from(
+        modulesWithDuplicatedKeysSet,
+      );
+      this.logger.warn('Found duplicated vetted keys', {
+        blockHash: blockData.blockHash,
+        duplicatedKeys: Array.from(duplicatedKeys),
+        moduleAddressesWithDuplicates: moduleAddressesWithDuplicatesList,
+      });
 
-    return {
-      nonce,
-      unusedKeys: keys.map((srKey) => srKey.key),
-      isDepositsPaused,
-      stakingModuleId: stakingRouterModule.id,
-      blockHash,
-    };
+      this.guardianMetricsService.incrDuplicatedVettedUnusedKeysEventCounter();
+      return moduleAddressesWithDuplicatesList;
+    }
+
+    return [];
+  }
+
+  private hasDuplicateKeys(stakingModules: Map<number, number>): boolean {
+    const moduleCounts = Array.from(stakingModules.values());
+
+    return stakingModules.size > 1 || moduleCounts[0] > 1;
+  }
+
+  public excludeModulesWithDuplicatedKeys(
```

Comment:
Looks like a simple filter that removes modules by id
`modules.filter((module)=>duplicatedModules.indexOf(stakingModuleId) != -1)`
Why so much code)

Reply:
it is enough  return stakingModulesData.filter(
        ({ stakingModuleId }) =>
          !modulesIdWithDuplicateKeys.includes(stakingModuleId),
      );

Reply:
fixed